salesagility/SuiteCRM
# Fixed #7740 - Discussion on replacing the StateChecker with database truncation in tests

## Description



WIP PR for discussing/investigating alternatives to StateChecker/StateSaver. The goal of this change is to reduce instances of "flaky tests" while also reducing the complexity involved when writing unit tests (See #7740 for details). 

I've so far implemented two traits for truncating DB tables and for SQL transactions and rollbacks that will run at the start of every test. From my testing, it seems that truncating the DB would significantly increase the test time which may be a good solution for smaller projects but I don't think it would be viable here. I would instead propose we use transactions/rollbacks for DB operations while global variables can be handled by [PHPUnit](https://thephp.cc/news/2015/04/phpunit-4-6-and-global-state). We'll potentially need to implement a custom handler for file operations. 

## Motivation and Context

Issue reference: #7740 

## How To Test This

1. Check Travis
2. Run unit tests locally: ```./vendor/bin/phpunit --stop-on-failure --stop-on-error --colors --configuration ./tests/phpunit.xml.dist ./tests/unit/phpunit```

## Types of changes


### Final checklist


- [X] My code follows the code style of this project found [here](htps://docs.suitecrm.com/community/contributing-code/coding-standards/).
